src/machine/board_arduino.go

```diff
@@ -43,3 +43,11 @@ const (
 	UART_TX_PIN Pin = PD1
 	UART_RX_PIN Pin = PD0
 )
+
+// SPI pins
+const (
+	SPI0_SCK_PIN Pin = PB5
+	SPI0_SDO_PIN Pin = PB3
+	SPI0_SDI_PIN Pin = PB4
+	SPI0_CS_PIN  Pin = PB2
```

Comment:
I see that some other boards also have a defined CS pin, but to be honest I don't think it should be used like this. The whole point of the chip select pin is to select which (of multiple) external devices are currently in use, and therefore you should use one per external device, not one per SPI peripheral.

```suggestion
```

Reply:
I'm fine removing the CS pin.

I initially went with the various documents from Atmel (all have a CS or SS pin defined even though they're just GPIO).  In the past AVR32 or PIC chips I've seen have had a special group of CS pins so that the SPI hardware could do the enable/disable automatically based on which chip you wanted to send/receive from ... all via interrupts.  From what I can tell, that pattern doesn't exist in any of the machines tinygo supports.

Reply:
I moved the implementation up to atmega since from what I could tell all the atmega chips have the same SPI hardware register implementation.

I agree with the pins being defined per chip or board or by the user.  It's just really nice to be able to wire & when there's only one possible choice it "just works".  This seems consistent with I2C and USART as well.

Reply:
> In the past AVR32 or PIC chips I've seen have had a special group of CS pins so that the SPI hardware could do the enable/disable automatically based on which chip you wanted to send/receive from ... all via interrupts. From what I can tell, that pattern doesn't exist in any of the machines tinygo supports.

Yeah, I'm aware this exists. We might want to add support for that eventually, but would need to find a way to express it in a portable way. And in any case, it only makes sense once DMA is supported (which it currently isn't, for the most part).

> I moved the implementation up to atmega since from what I could tell all the atmega chips have the same SPI hardware register implementation.

No. The atmega1284p for example has 3 SPI peripherals, so it can't use the exact same registers: there need to be 3 sets of them with different names. Maybe it has the same names for SPI0 but it can't have the same names for SPI1 and SPI2.

Reply:
I must be missing something, but I don't see the other 2 SPI peripherals in the datasheet.  I see the USART as SPI (many of the chips have that), but I didn't try to address those other 2 modes.

Reply:
I found a different supported chip that has the pattern you mentioned. `atmega1284p`

My larger question is would you rather have a bunch of duplicated code for each processor or have a more generic solution that works for all the Atmega chips?

Reply:
I found a different chip that matches your example: `atmega384pb`

My question is would you rather have a more generic implementation that all the Atmega chips can use, or do you want the code duplicated for each chip?

Reply:
I'm referring to this product page: https://www.microchip.com/wwwproducts/en/ATMEGA1284P

But maybe you're right, with a quick scan I can't find other SPI buses. Which kind of makes sense, as you can't really do much with multiple SPI peripherals if you don't also use DMA (to work with them in parallel).

Reply:
> My question is would you rather have a more generic implementation that all the Atmega chips can use, or do you want the code duplicated for each chip?

Ideally a generic solution if it is possible, but maybe it is not possible. For example, when I was working on a refactor of PWM support I couldn't really come up with a working generic solution.